src/daemon/docker_exec.sh

```diff
@@ -24,41 +24,121 @@ function trap_error {
       echo "export $value" >> /root/.bashrc
     done
 
-    sleep 365d
+    sleep infinity
   else
     # If NOTRAP is defined, we need to return true to avoid triggering an ERR
     true
   fi
 }
 
 child_for_exec=1
-function _term {
-  echo "Sending SIGTERM to PID $child_for_exec"
+
+function teardown {
+  # Disabling the traps to avoid a cascading
+  trap - SIGINT SIGILL SIGABRT SIGFPE SIGSEGV SIGTERM SIGBUS SIGCHLD SIGKILL
+
+  # This function is called when we got a signal reporting something died
+  # It will check the child_for_exec process exited
+  # Then it will execute the optional sigterm_cleanup_post and return the passed exit code
+  local signal_name=$1
+  local exit_code=$2
+  echo "teardown: managing teardown after $signal_name"
 
   # Disabling the ERR trap before killing the process
   # That's an expected failure so don't handle it
   # Doing "trap ERR" or "trap - ERR" didn't worked :/
   NOTRAP="yes"
-  declare -F sigterm_cleanup_pre && sigterm_cleanup_pre
-  kill -TERM "$child_for_exec" 2>/dev/null
-  declare -F sigterm_cleanup_post && sigterm_cleanup_post
+
+  # If we receive SIGTERM, it means the process is supposed to still be alive
+  # So let's call sigterm_cleanup_pre if any defined
+  if [ "$signal_name" = "SIGTERM" ]; then
+    declare -F sigterm_cleanup_pre && sigterm_cleanup_pre
+  fi
+
+  # Sending the TERM signal to the exec'd program
+  # If we got a SIGTERM, that is propagating the signal,
+  # If we got any other signal, we need to teardown as the stability is no more guaranteed
+  [ -e /proc/$child_for_exec ] && (echo "teardown: Sending SIGTERM to PID $child_for_exec"; kill -TERM "$child_for_exec" 2>/dev/null)
+
+  echo -n "teardown: Waiting PID $child_for_exec to terminate "
+  while [ -e /proc/$child_for_exec ]; do
```

Comment:
no timeout?

Reply:
I let a comment to explain why

Reply:
I finally change the code for more robustness